lib/socks_sspi.c

```diff
@@ -514,22 +507,33 @@ CURLcode Curl_SOCKS5_gssapi_negotiate(struct Curl_cfilter *cf,
     status = Curl_pSecFn->DecryptMessage(&sspi_context, &wrap_desc,
                                          0, &qop);
 
+    /* since sspi_w_token[1].pvBuffer is allocated by the SSPI in this case, it
+       must be freed in this block using FreeContextBuffer() instead of
+       potentially in error cleanup using free(). */
+
     if(check_sspi_err(data, status, "DecryptMessage")) {
       failf(data, "Failed to query security context attributes.");
+      Curl_pSecFn->FreeContextBuffer(sspi_w_token[1].pvBuffer);
+      sspi_w_token[1].pvBuffer = NULL;
```

Comment:
The point of the `goto error` is to have all the cleanup calls there, to avoid having them for every case we catch the errors. So shouldn't this FreeContextBuffer() call be moved to below `error:` instead?

Reply:
So in this particular case it is noted in the comment above. This is a special case because sspi_w_token[1].pvBuffer was allocated by the SSPI so it must be freed by the SSPI call instead of how it would normally be freed (by free). Also, the reason there's a lot of frees in the code is because those buffers are then reused later on. The goto error only handles the error case it doesn't do cleanup on success.